### PR TITLE
Add ability to reference waypoint.hcl variables between eachother

### PR DESCRIPTION
**Pull request by evanphx**
This change adds the ability refer to other waypoint.hcl defined variables using HCL templating. Additionally, it adds a new class of variable called internal variables. These variables are not exposed to the application and are designed to be composed into other variables.

This functionally fully supports completing dynamic variables into non-dynamic ones, allowing for a better where internal dynamic variables are composed into a env variables:

```hcl
config {
  internal = {
    dbhost = configdynamic("terraform-cloud", {
       organization = "foocorp",
       workspace = "production",
       output = "dbhost"
    })
  }

  env = {
    DATABASE_URL = "postgresql://${config.internal.dbhost}"
  }
}
```

### CODE REVIEW

File: internal/appconfig/watcher.go [hunk 1]
```diff
@@ -493,25 +497,46 @@ func configVarSortFunc(vars []*pb.ConfigVar) func(i, j int) bool {
 	}
 }
 
+// Used to just in tracking from the config split, through eval, and back
+// to exporting.
```
Comment (mitchellh) [inline internal/appconfig/watcher.go:501]: nitpick but would explain what static vs dynamic is here at a high level here and in the comment above `dynamicVar`

File: internal/appconfig/watcher.go [hunk 2]
```diff
@@ -662,14 +687,32 @@ func buildAppConfig(
...
 	if len(dynamic) == 0 {
-		return static
+		return expandStaticVars(log, &ectx, staticVars)
 	}
 
+	// The way this next bit works is that any static values that referenced
+	// other static values have already been expanded before they make it this far,
+	// which means that if a static variable still contains an HCL template, it's
+	// going to reference a dynamic variable. And because dynamic variables can't
+	// reference other variables, the job is pretty easy.
```
Comment (mitchellh) [inline internal/appconfig/watcher.go:701]: Took me a sec to consider this but you're of course absolutely right... since dynamic vars can't reference anything by their nature (just fundamentally), this simplifies things a LOT. This is a really nice constraint that I didn't consider before! 👍

File: internal/config/app_config_sort.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+// Captures the information about a variable calculated during sort to be used
+// without recalculation by the caller.
+type analyzedPair struct {
+	Pair     hcl.KeyValuePair
+	Name     string
+	Internal bool
+	Refs     []string
+}
+
+type VariableLoopError struct {
```
Comment (mitchellh) [inline internal/config/app_config_sort.go:36]: Comment on this.

Second, since its exported, I would consider putting it towards the top of the file so that people browsing the file may see it sooner but thats purely a matter of personal taste.

File: internal/config/app_config_sort.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func (v *VariableLoopError) Error() string {
+	return fmt.Sprintf("loop detected amongst variables: %s", strings.Join(v.LoopVars, ", "))
+}
+
+// sortVars performs a topological sort on EnvRaw and InternalRaw. The sort
+// yields the pairs in most referenced to least referenced order. Meaning
+// that the if pair X references pair R, then R will be before X in the slice.
+func (c *genericConfig) sortVars(ctx *hcl.EvalContext) ([]*analyzedPair, error) {
```
Comment (mitchellh) [inline internal/config/app_config_sort.go:47]: I think this needs unit tests. Given its a pure function that is executing a well known algorithm, I think we can cover all the edge cases and ensure this really works. I know its tested incidentally through higher level tests, but they're mostly happy path cases.

**Review by mitchellh — CHANGES_REQUESTED** [review]
This looks really good! Relatively simple overall, at least compared to what I thought would be required.

I think the one **big** question I have is how escaping works in this scenario (i.e. I literally want `${foo}` in my value). This might just be handled by `internal/pkg/partial`. If so, I don't see any tests for it and I think we should definitely cover escaping tests at most levels.

File: internal/appconfig/watcher.go (hunk 1 above)
Comment (krantzinator) [reply internal/appconfig/watcher.go:501]: also nitpick: confusing wording `Used to just in tracking`

Comment (evanphx) [reply internal/appconfig/watcher.go:511]: Both fixed!

File: internal/appconfig/watcher.go (hunk 2 above)
Comment (evanphx) [reply internal/appconfig/watcher.go:711]: I realized it as I started to code it up and did a little fist pump in the air.

File: internal/config/app_config_sort.go (hunk 3 above)
Comment (evanphx) [reply internal/config/app_config_sort.go:38]: Done!

File: internal/config/app_config_sort.go (hunk 4 above)
Comment (evanphx) [reply internal/config/app_config_sort.go:49]: Yup, added a few. There aren't a ton of edge cases so it should cover them.

File: internal/config/app_config_sort.go [hunk 5]
```diff
@@ -63,6 +65,10 @@ func (c *genericConfig) sortVars(ctx *hcl.EvalContext) ([]*analyzedPair, error)
 	}
 
 	for _, m := range maps {
+		if m.expr == nil {
```
Comment (mitchellh) [inline internal/config/app_config_sort.go:68]: Quick comment: why/when does this happen?

File: internal/config/app_test.go [hunk 6]
```diff
@@ -201,6 +201,23 @@ func TestConfigApp_compare(t *testing.T) {
...
+				require := require.New(t)
+
+				vars, err := c.Config.ConfigVars()
+				require.NoError(err)
+
+				// test the static value
+				require.Len(vars, 3)
+				static, ok := vars[2].Value.(*pb.ConfigVar_Static)
+				require.True(ok)
+				require.Equal("lower(config.internal.greeting, \"FOO\")", static.Static)
```
[the remark this answers is not in the document]
Comment (evanphx) [reply internal/config/app_test.go]: No prob! fixed.

File: internal/config/app_config.go [hunk 7]
```diff
@@ -36,6 +37,8 @@ func (c *genericConfig) ConfigVars() ([]*pb.ConfigVar, error) {
 	return c.envVars()
 }
 
+var hclEscaper = strings.NewReplacer("${", "$${", "%{", "%%{")
```
[the remark this answers is not in the document]
Comment (evanphx) [reply internal/config/app_config.go:40]: It's nifty!

File: internal/config/app_config_sort.go (hunk 5 above)
Comment (evanphx) [reply internal/config/app_config_sort.go]: Cleanup the logic, it was only for a test but I just modified the tests to not require it.

File: .changelog/1382.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
+config: Add ability to define internal variables and compose variables together via templating
```
Comment (mitchellh) [inline .changelog/1382.txt:2]: Nitpick, we may want to turn this one into a bold heading style (see 0.3 or 0.2 or w/e changelog) but we can do that human-powered when we do a major release?

**Review by mitchellh — APPROVED** [review]

File: .changelog/1382.txt (hunk 8 above)
Comment (briancain) [reply .changelog/1382.txt:2]: You should be able to use markdown here and go-changelog will respect it, so adding the `*` _should_ be fine to do here if we want to bold it

Comment (mitchellh) [reply .changelog/1382.txt:2]: Cool, yeah worth considering, I think this is a big feature.

File: internal/ceb/config.go [hunk 9]
```diff
@@ -193,6 +195,58 @@ func (ceb *CEB) watchConfig(
...
+	"HUP":    unix.SIGHUP,
+	"INT":    unix.SIGINT,
+	"IO":     unix.SIGIO,
+	"KILL":   unix.SIGKILL,
+	"PIPE":   unix.SIGPIPE,
+	"PROF":   unix.SIGPROF,
+	"QUIT":   unix.SIGQUIT,
+	"SEGV":   unix.SIGSEGV,
+	"STOP":   unix.SIGSTOP,
+	"SYS ":   unix.SIGSYS,
```
Comment (briancain) [inline internal/ceb/config.go:238]: I saw this was fixed on the other branch but maybe we can fix this here too since it's ready for merge
```suggestion
	"SYS":   unix.SIGSYS,
```

**Review by briancain — APPROVED** [review]
